contributors/design-proposals/storage/dynamic_volume_limit.md

```diff
@@ -0,0 +1,119 @@
+# Dynamic attached volume limits
+
+## Goals
+
+Currently the number of volumes attachable to a node is either hard-coded or only configurable via an environment variable. Also
+existing limits only apply to well known volume types like EBS, GCE and is not available to all volume plugins.
+
+This proposal enables any volume plugin to specify those limits and also allows same volume type to have different volume
+limits depending on type of node.
+
+## Implementation Design
+
+### Prerequisite
+
+* This feature will be protected by an alpha feature gate, so as API and CLI changes needed for it. We are planning to call
+  the feature `DynamicVolumeThreshold`.
+
+### API Changes
+
+There is no API change needed for this feature. However existing `node.Status.Capacity` and `node.Status.Allocatable` will
+be extended to cover volume limits available on the node too.
+
+The key name that will store volume will be start with prefix `storage-limits-`. The volume limit key will respect
+format restrictions applied to Kubernetes Resource names. Volume limit key for existing plugins might look like:
+
+
+* `storage-limits-aws-ebs`
+* `storage-limits-gce-pd`
+* `storage-limits-csi-<driver_name>`
```

Comment:
what is the max length of a CSI driver name? does that name, prefixed with `storage-limits-csi-` (or whatever prefix is chosen) fit within the resourcename length restrictions?

Reply:
hard-coding the attach resource name for in-tree volume types seems ok short-term, but for CSI volumes, I expected to see a representation in the PV CSIPersistentVolumeSource of what resource consumption was associated with the PV, among which could be the volume count resource.

that lets the scheduler avoid needing to call the volume plugin manager:
* in-tree volume types have hard-coded `f(pod) -> resources` and `f(pv) -> resources` transforms
* CSI volumes have `f(pv) -> resources` represented in the CSIPersistentVolumeSource API

then the scheduler can use normal resource fit calculations to see if a given pod and its PVs fit on a given node

Reply:
I have tried addressing it here - https://github.com/kubernetes/kubernetes/pull/64154#discussion_r190955132